### PR TITLE
Post authentication creates user in user catalogue for new logins

### PR DESCRIPTION
**Pull request by janbj**

### CODE REVIEW

File: cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java [hunk 1]
```diff
@@ -59,17 +60,33 @@ public UserApiClient(HttpClient httpClient, ObjectMapper objectMapper, Environme
 
     @Override
     @JacocoGenerated
-    public void createUser(User user) {
+    public Optional<User> createUser(User user) {
+        logger.info("Requesting user creation for username: " + user.getUsername());
+        return createNewUser(user)
+            .stream()
+            .filter(responseIsSuccessful())
```
Comment (axthosarouris) [inline cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java:67]: refactor responseIsSuccessful to the following and call the filter like `filter(this::responseIsSuccessful)`

```
 private boolean responseIsSuccessful(HttpResponse<String> response) {
        return response.statusCode() == HttpStatus.SC_OK;
    }
```

File: cognito-post-authentication/src/test/java/no/unit/nva/cognito/PostAuthenticationHandlerTest.java [hunk 2]
```diff
@@ -97,18 +100,38 @@ public void handleRequestReturnsErrorWhenCustomerIsNotFound() {
         Assertions.assertEquals(NOT_FOUND_ERROR_MESSAGE + SAMPLE_ORG_NUMBER, exception.getMessage());
     }
 
+    @Test
+    public void handleRequestReturnsErrorWhenCustomerIsCreated() {
```
Comment (axthosarouris) [inline cognito-post-authentication/src/test/java/no/unit/nva/cognito/PostAuthenticationHandlerTest.java:104]: Is customer an institution?

Comment (janbj) [reply cognito-post-authentication/src/test/java/no/unit/nva/cognito/PostAuthenticationHandlerTest.java:104]: Yes, customer institution.

File: cognito-post-authentication/src/main/java/no/unit/nva/cognito/PostAuthenticationHandler.java [hunk 3]
```diff
@@ -74,19 +74,10 @@ public Event handleRequest(Event event, Context context) {
         User user = getUserFromCatalogueOrAddUser(userAttributes);
 
         updateUserDetailsInUserPool(userPoolId, userName, userAttributes, user);
-        updateUserGroupsInUserPool(userPoolId, userName, user);
 
         return event;
     }
 
-    private void updateUserGroupsInUserPool(String userPoolId, String userName, User user) {
```
Comment (janbj) [inline cognito-post-authentication/src/main/java/no/unit/nva/cognito/PostAuthenticationHandler.java:82]: Removed this behavior as we currently have no reason to set groups.

File: cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApi.java [hunk 4]
```diff
@@ -7,5 +7,5 @@
 
     Optional<User> getUser(String username);
 
-    void createUser(User user);
+    Optional<User> createUser(User user);
```
Comment (axthosarouris) [inline cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApi.java:10]: Does an empty Optional mean failure? If yes, then failure should be indicated with an Exception. If not, what is the step after the user has not been created? You do not use the Optional in your production code. only in tests

File: cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java [hunk 5]
```diff
@@ -59,17 +59,33 @@ public UserApiClient(HttpClient httpClient, ObjectMapper objectMapper, Environme
...
-    public void createUser(User user) {
+    public Optional<User> createUser(User user) {
+        logger.info("Requesting user creation for username: " + user.getUsername());
+        return createNewUser(user)
+            .stream()
+            .filter(this::responseIsSuccessful)
+            .map(this::tryParsingUser)
+            .collect(SingletonCollector.tryCollect())
+            .flatMap(this::flattenNestedAttempts)
+            .toOptional(this::logErrorParsingUserInformation);
```
Comment (axthosarouris) [inline cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java:70]:
```suggestion

orElseThrow(failure->logErrorAndReturnException(failure)
....

private void logErrorAndReturnException(failure){
 logger.error(failure.getException.getMessage(), failure.getException);
return new CreateUserFailedException();
}
```

File: cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java [hunk 6]
```diff
@@ -59,17 +59,33 @@ public UserApiClient(HttpClient httpClient, ObjectMapper objectMapper, Environme
...
 
-    private Optional<HttpResponse<String>> fetchUserInformation(String orgNumber) {
-        return attempt(() -> formUri(orgNumber))
+    private Optional<HttpResponse<String>> createNewUser(User user) {
+        return attempt(() -> formUri())
             .map(URIBuilder::build)
-            .map(this::buildHttpRequest)
+            .map(uri -> buildCreateUserRequest(uri, user))
+            .map(this::sendHttpRequest)
+            .toOptional(failure -> logResponseError(failure));
```
Comment (axthosarouris) [inline cognito-post-authentication/src/main/java/no/unit/nva/cognito/service/UserApiClient.java:82]: maybe here as well you can use an Exception.

**Pull request merged**